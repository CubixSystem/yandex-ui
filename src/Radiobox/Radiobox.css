.Radiobox {
    display: inline-block;

    font-family: var(--radiobox-font-family);
    user-select: none;

    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
}

.Radiobox-Radio_disabled {
    pointer-events: none;
}
