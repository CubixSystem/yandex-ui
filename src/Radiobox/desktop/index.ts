export * from '../Radiobox@desktop';
export * from '../_size/Radiobox_size_m';
export * from '../_size/Radiobox_size_s';
export * from '../_theme/Radiobox_theme_normal@desktop';
export * from '../_theme/Radiobox_theme_pseudo@desktop';
export * from '../_view/Radiobox_view_default@desktop';
export * from '../_view/Radiobox_view_outline@desktop';
