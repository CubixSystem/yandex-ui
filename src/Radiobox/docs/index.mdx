---
id: Radiobox
path: /components/radiobox/examples

title: Radiobox
description: Компонент для создания радиопереключателя.

links:
 - label: source
   url: https://github.yandex-team.ru/search-interfaces/frontend/tree/master/packages/lego-components/src/Radiobox
 - label: startrek
   url: https://search.yandex-team.ru/stsearch?text=Radiobox.ts&facet.queue=ISL&facet.type=bug&facet.status=128
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/L-E-G-O?node-id=0%3A1
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fcontrols-radiobox-desktop--playground

tabs:
 - Examples
 - Code
---
