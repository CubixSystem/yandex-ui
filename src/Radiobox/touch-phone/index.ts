export * from '../Radiobox@touch-phone';
export * from '../_size/Radiobox_size_m';
export * from '../_size/Radiobox_size_s';
export * from '../_theme/Radiobox_theme_normal';
export * from '../_theme/Radiobox_theme_pseudo';
export * from '../_view/Radiobox_view_default';
export * from '../_view/Radiobox_view_outline';
