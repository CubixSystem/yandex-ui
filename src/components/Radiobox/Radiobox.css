.Radiobox {
    display: inline-block;

    font-family: var(--control-font-family);
    user-select: none;

    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
}

.Radiobox-Radio_disabled {
    pointer-events: none;
}
