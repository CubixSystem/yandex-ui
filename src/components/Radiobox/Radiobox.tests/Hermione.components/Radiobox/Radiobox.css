.Radiobox-Radio {
    margin-right: 14px;
}
