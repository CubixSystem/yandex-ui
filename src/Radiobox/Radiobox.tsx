import React, { FC, Ref, ReactNode, useState, useEffect, ChangeEventHandler } from 'react';
import { useComponentRegistry } from '@bem-react/di';
import { cn } from '@bem-react/classname';

import { IRadioboxRegistry } from './Radiobox.registry';
import { RadioGroupProvider } from './Radiobox.context';
import { RadioboxRadioProps } from './Radio/Radiobox-Radio';
import './Radiobox.css';

export const cnRadiobox = cn('Radiobox');

// Для options используем label вместо children, а так же удаляем
// checked и onChange, т.к. они предполагают использование `Radio` элемента.
type PartialRadioProps = Omit<RadioboxRadioProps, 'children' | 'checked' | 'onChange'>;

export type RadioOptionProps = PartialRadioProps & {
    /**
     * Текст подписи к переключателю
     */
    label: ReactNode;
};

export type RadioboxProps = {
    /**
     * Набор переключателей
     *
     * Может быть использовано вместо `children`
     *
     * @default []
     */
    options?: RadioOptionProps[];

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: Ref<HTMLSpanElement>;

    /**
     * Неактивное состояние всей группы переключателей
     */
    disabled?: boolean;

    /**
     * Набор переключателей c использованием `Radio` элемента
     *
     * Может быть использовано вместо `options`
     */
    children?: ReactNode;

    /**
     * Дополнительный класс у корневого DOM-элемента
     */
    className?: string;

    /**
     * Текущее выбранное значение в группе
     */
    value?: string;

    /**
     * Имя для всех `Radio` элементов
     */
    name?: string;

    /**
     * Коллбек, который срабатывает при изменении значения
     */
    onChange?: ChangeEventHandler<HTMLInputElement>;
};

/**
 * Компонент для создания радиопереключателя.
 * @param {RadioboxProps} props Свойства компонента.
 */
export const Radiobox: FC<RadioboxProps> = ({
    children,
    className,
    disabled,
    innerRef,
    name,
    onChange,
    options = [],
    value,
    // Извлекаем свойства, т.к. они не нужны на DOM узле
    // FIXME: https://github.com/bem/bem-react/issues/381
    // @ts-ignore
    view: _view,
    // @ts-ignore
    size: _size,
    // @ts-ignore
    theme: _theme,
}) => {
    const { Radio } = useComponentRegistry<IRadioboxRegistry>(cnRadiobox());
    const [context, setContext] = useState({ onChange, value, disabled, name });

    useEffect(() => {
        setContext({ onChange, value, disabled, name });
    }, [onChange, value, disabled, name]);

    return (
        <span className={cnRadiobox(null, [className])} ref={innerRef} role="radiogroup">
            {/* prettier-ignore */}
            <RadioGroupProvider value={context}>
                {children || options.map(({ label, ...option }) => (
                    <Radio key={option.value} {...option}>
                        {label}
                    </Radio>
                ))}
            </RadioGroupProvider>
        </span>
    );
};

Radiobox.displayName = cnRadiobox();
