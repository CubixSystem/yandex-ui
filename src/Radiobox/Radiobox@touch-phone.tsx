import { withRegistry } from '@bem-react/di';

import { Radiobox as RadioboxCommon } from './Radiobox';
import { radioboxRegistry } from './Radiobox.registry/touch-phone';

export * from './Radiobox';
export { RadioboxRadioProps as RadioProps, RadioboxRadio as Radio } from './Radio/Radiobox-Radio';
export const Radiobox = withRegistry(radioboxRegistry)(RadioboxCommon);
