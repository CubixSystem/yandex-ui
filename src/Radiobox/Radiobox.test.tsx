import React, { createRef, ComponentType } from 'react';
import { mount, render } from 'enzyme';
import { ExtractProps } from '@bem-react/core';

import { Radiobox as RadioboxCommon } from './Radiobox';
import { Radiobox as RadioboxDesktop } from './Radiobox@desktop';
import { Radiobox as RadioboxTouchPad } from './Radiobox@touch-pad';
import { Radiobox as RadioboxTouchPhone } from './Radiobox@touch-phone';
import { serverEnvironmentSetup, delay } from '../internal/utils';

// prettier-ignore
const platforms = [
    ['desktop', RadioboxDesktop],
    ['touch-pad', RadioboxTouchPad],
    ['touch-phone', RadioboxTouchPhone],
];

type RadioboxProps = ExtractProps<typeof RadioboxDesktop>;

// prettier-ignore
const options = [
    { label: 'Option A', value: 'a' },
    { label: 'Option B', value: 'b' },
];

describe.each<any>(platforms)('Radiobox@%s', (_platform, Radiobox: ComponentType<RadioboxProps>) => {
    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(
                render(<Radiobox onChange={() => null} value="a" options={options} className="mix" name="radio" />),
            ).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(
                render(<Radiobox onChange={() => null} value="a" options={options} className="mix" name="radio" />),
            ).toMatchSnapshot();
        });

        test('должен быть установлен корректный displayName', () => {
            expect(RadioboxCommon.displayName).toBe('Radiobox');
        });

        test('должен устанавливать ссылку на корневой DOM элемент', async () => {
            const innerRef = createRef<HTMLElement>();
            mount(<Radiobox onChange={() => null} value="" options={[]} innerRef={innerRef} />);
            await delay(100);
            expect(innerRef.current).not.toBe(null);
        });

        test('должен вызывать событие onChange с event при клике на контрол', async () => {
            const onChange = jest.fn();

            const wrapper = mount(<Radiobox value="a" options={options} onChange={onChange} />);
            wrapper
                .find('input')
                .at(1)
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledTimes(1);
            expect(onChange.mock.calls[0][0].target.checked).toBe(true);
        });

        test('должен устанавливать checked/aria-checked при обновлении свойства value', () => {
            const wrapper = mount(<Radiobox onChange={() => null} options={options} value="a" />);
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('checked'),
            ).toBe(true);
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('aria-checked'),
            ).toBe(true);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('checked'),
            ).toBe(false);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('aria-checked'),
            ).toBe(false);
            wrapper.setProps({ value: 'b' });
            wrapper.update();
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('checked'),
            ).toBe(false);
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('aria-checked'),
            ).toBe(false);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('checked'),
            ).toBe(true);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('aria-checked'),
            ).toBe(true);
        });

        test('должен добавлять и удалять атрибут disabled при обновлении через props', () => {
            const wrapper = mount(<Radiobox onChange={() => null} value="a" options={options} />);
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('disabled'),
            ).toBe(undefined);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('disabled'),
            ).toBe(undefined);
            wrapper.setProps({ disabled: true });
            wrapper.update();
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('disabled'),
            ).toBe(true);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('disabled'),
            ).toBe(true);
        });

        test('должен позволять блокировать и разблокировать отдельные опции через props', () => {
            const wrapper = mount(<Radiobox onChange={() => null} value="a" options={options} />);
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('disabled'),
            ).toBe(undefined);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('disabled'),
            ).toBe(undefined);
            const optionsWithBDisabled = [
                { label: 'Option A', value: 'a' },
                { label: 'Option B', value: 'b', disabled: true },
            ];
            wrapper.setProps({ options: optionsWithBDisabled });
            wrapper.update();
            expect(
                wrapper
                    .find('input')
                    .at(0)
                    .prop('disabled'),
            ).toBe(undefined);
            expect(
                wrapper
                    .find('input')
                    .at(1)
                    .prop('disabled'),
            ).toBe(true);
        });
    });
});
