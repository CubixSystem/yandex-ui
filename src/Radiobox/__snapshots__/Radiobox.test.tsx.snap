// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radiobox@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;

exports[`Radiobox@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Radiobox mix"
  role="radiogroup"
>
  <label
    class="Radiobox-Radio Radiobox-Radio_checked"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="true"
        aria-labelledby="label-0"
        autocomplete="off"
        checked=""
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="a"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option A
    </span>
  </label>
  <label
    class="Radiobox-Radio"
  >
    <span
      class="Radiobox-Box"
    >
      <input
        aria-checked="false"
        aria-labelledby="label-0"
        autocomplete="off"
        class="Radiobox-Control"
        name="radio"
        type="radio"
        value="b"
      />
    </span>
    <span
      aria-hidden="true"
      class="Radiobox-Text"
      id="label-0"
    >
      Option B
    </span>
  </label>
</span>
`;
